File: auth-for-mcp/fastmcp-mcp-customtokenexchange-python/src/tools.py
import json
import logging

import httpx
from mcp.server.fastmcp import Context

from .auth0 import Auth0Mcp
from .auth0.authz import register_required_scopes, require_scopes
from .config import get_config

logger = logging.getLogger(__name__)


def register_tools(auth0_mcp: Auth0Mcp) -> None:
    """
    Register all tools with the MCP server.
    """
    mcp = auth0_mcp.mcp
    config = get_config()

    async def exchange_custom_token(api_client, subject_token: str) -> dict:
        """Exchange subject token for access token via Custom Token Exchange."""
        result = await api_client.get_token_by_exchange_profile(
            subject_token=subject_token,
            subject_token_type=config.mcp_auth0_subject_token_type,
            audience=config.api_auth0_audience,
            scope=config.mcp_auth0_exchange_scope or None
        )
        return {"token": result["access_token"], "scopes": result.get("scope", "")}

    # A MCP tool with required scopes
    @mcp.tool(
        name="whoami",
        title="Who Am I Tool",
        description="Returns information about the authenticated user",
        annotations={"readOnlyHint": True}
    )
    @require_scopes(["tool:whoami"])
    async def whoami(ctx: Context) -> str:
        auth_info = ctx.request_context.request.state.auth

        response_data = {
            "user": auth_info.get("extra", {}),
            "scopes": auth_info.get("scopes", []),
        }
        return json.dumps(response_data, indent=2)

    # A MCP tool with Custom Token Exchange
    @mcp.tool(
        name="greet",
        title="Greet User with Custom Token Exchange",
        description="Greet a user with personalized authentication information retrieved from an upstream API using Custom Token Exchange.",
        annotations={"readOnlyHint": True}
    )
    @require_scopes(["tool:greet"])
    async def greet(name: str, ctx: Context) -> str:
        user_name = name.strip() if name else "there"
        auth_info = ctx.request_context.request.state.auth
        user_id = auth_info.get("extra", {}).get("sub")

        logger.info(f"Greet tool invoked for user: {user_id}")

        # Exchange token and call upstream API
        exchange_result = await exchange_custom_token(
            ctx.request_context.request.state.api_client,
            auth_info["token"]
        )

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{config.api_base_url}/api/private-scope",
                headers={"authorization": f"Bearer {exchange_result['token']}"}
            )
            upstream_result = response.json()
            logger.info(f"Upstream API response: {upstream_result}")

        return f"""Hello, {user_name} ({user_id})!
Upstream API Response: {json.dumps(upstream_result, indent=2)}"""

    # Register all scopes used by tools for Protected Resource Metadata
    register_required_scopes(auth0_mcp)

File: auth-for-mcp/fastmcp-mcp-customtokenexchange-js/src/tools.ts
import { z } from "zod";
import { FastMCP } from "fastmcp";
import { FastMCPAuthSession } from "./types.js";
import { exchangeCustomToken } from "./auth0.js";

export const MCP_TOOL_SCOPES = ["tool:greet", "tool:whoami"];

function hasAllScopes(
  requiredScopes: readonly string[]
): (auth: FastMCPAuthSession) => boolean {
  return (auth: FastMCPAuthSession) => {
    const userScopes = auth.scopes;
    return requiredScopes.every((scope) => userScopes.includes(scope));
  };
}

async function bearerForUpstream(subjectToken: string) {
  if (!subjectToken) return { token: null, scopes: null };

  try {
    const result = await exchangeCustomToken(subjectToken);
    return {
      token: result.accessToken,
      scopes: result.scope,
    }
  } catch (err) {
    console.error('Error during token exchange:', err);
    throw err;
  }
}

export function registerTools(mcpServer: FastMCP<FastMCPAuthSession>) {
  mcpServer.addTool({
    name: "whoami",
    description: "Returns information about the authenticated user",
    annotations: {
      title: "Who Am I? (FastMCP)",
      readOnlyHint: true,
    },
    canAccess: hasAllScopes(["tool:whoami"]),
    execute: async (_args, { session: authInfo }) => {
      const info = { user: authInfo?.extra, scopes: authInfo?.scopes };
      return JSON.stringify(info, null, 2);
    },
  });

  mcpServer.addTool({
    name: "greet",
    description:
      "Greet a user with personalized authentication information retrieved from an upstream API using Custom Token Exchange.",
    annotations: {
      title: "Greet User with Custom Token Exchange",
      readOnlyHint: true,
    },
    parameters: z.object({
      name: z
        .string()
        .optional()
        .describe("The name of the person to greet (optional)."),
    }),
    canAccess: hasAllScopes(["tool:greet"]),
    execute: async (args, { session: authInfo }) => {
      const { name } = args;
      const userName = name ?? "there";

      console.log(`Greet tool invoked for user: ${authInfo?.extra?.sub}`);

      // Perform Custom Token Exchange and call upstream API
      const sourceToken = authInfo!.token;
      const { token: bearer, scopes } = await bearerForUpstream(sourceToken);
      const headers = { 'content-type': 'application/json', authorization: `Bearer ${bearer}` };

      // Call to an upstream API using the exchanged token
      const res = await fetch(`${process.env.API_BASE_URL}/api/private-scope`, {
        method: 'GET',
        headers,
      });
      const upstreamResult = await res.json();
      console.log('Upstream API response:', upstreamResult);

      return `
        Hello, ${userName} (${authInfo?.extra?.sub})!
        Upstream API Response: ${JSON.stringify(upstreamResult, null, 2)}
        `.trim();
    },
  });

  // This tool does not require any scopes
  mcpServer.addTool({
    name: "get_datetime",
    description: "Returns the current UTC date and time",
    annotations: {
      title: "Get DateTime",
      readOnlyHint: true,
    },
    execute: async () => {
      const utcDateTime = new Date().toISOString();
      return `Current UTC DateTime: ${utcDateTime}`;
    },
  });
}
